### ISSUE TITLE
Lack of randomness

### ISSUE THREAD

**Issue by Tim---** [robertdavidgraham/masscan#505]
Hi,

I noticed that the masscan randomness is not very random. I understand that perfect randomness and distribution of the IPs during the scan is not the main goal, but I think this can have two bad consequences:

- The scans are not uniform in the IP space. For example, if you scan a /10 at 1M pps, you expect each /20 to be hit at 1k pps. But in practice, some subnetworks will be hit a lot more than others at a given time. This means that you have to reduce your scan rate to avoid overwhelming the routers.
- The scans are not uniform in the port space. This can cause problem if you want to do a "sample" scan of a network to see which ports are more popular. After 10% of the scan, you might think that your sample will match the global distribution on the network. But in practice, some ports will be hit a lot more than others and will yield more results. You can feel it when you do a scan of 0.0.0.0/0 on all ports and you barely see any tcp/80 open at the start of the scan.

The graphics below shows the difference in randomness:
![difference3](https://user-images.githubusercontent.com/5160665/87064437-078a7080-c210-11ea-925c-9dd14cf667f4.png)

Basically, I simulated a scan of 0.0.0.0/0 using masscan, zmap and real randomness. After 100k probes, I wanted to see if each /10 got the same amount of probes.
We can see that zmap matches closely real randomness, and that each /10 got between 75 and 125 probes.
However, for masscan, each /10 will get between 0 and 300 probes. Some network are hit a lot more than others.

Before I dig further, I wanted to know if anyone already stumbled upon this problem ? If so, do you have any pointers or ideas of how this could be resolved ?

Regards.

**Comment by p-l-** [comment]
Hi,

FWIW, it reminds me of this paper: https://jhalderm.com/pub/papers/zmap10gig-woot14.pdf.

**Comment by Frky** [comment]
Hi,

This problem is interesting, and I think it would be useful to find its cause.
I, with the help of @p-l-, tried to reproduce the results you present, with the following command:

```
 sudo bin/masscan --router-mac 00:01:02:03:04:05 --adapter-ip 10.0.0.2 --range 0.0.0.0/24 --exclude 255.255.255.255 -p80 --rate 1000

```

I stopped the capture after 100k packets received to match your experimental setup, and I get something similar to what you have (only worse, but maybe it's due to experimental parameters?):

![veth_ipv4_100000](https://user-images.githubusercontent.com/3105926/87303315-4885cc00-c513-11ea-8ac7-bbdef68b8785.png)

In addition to that, I performed the same test, but calling directly the `blackrock` generator itself, in order to confirm that the lack of randomness is indeed coming from it and not because of the way its output is used. To do so, I used this code snippet (note that I also tested the Linear Congruential Generator, but it does not seem to be used anymore):

```
// command line: ./bin/masscan 4294967296 100000 2>/dev/null

int main(int argc, char *argv[])
{
    size_t N = atol(argv[1]);
    size_t nb = atol(argv[2]);
#if 0
    uint64_t a;
    uint64_t c;
    uint64_t x = 0;
    lcg_calculate_constants(N, &a, &c, 0);
    for (size_t i = 0; i < nb; i++) {
        x = lcg_rand(x, a, c, N);
        printf("%lu\t", x);
    }
    printf("\n");
#else
    struct BlackRock blackrock;
    srand(time(0));
    uint64_t seed = 0x86b5d2719623fc59;
    blackrock_init(&blackrock, N, seed, 4);
    uint64_t x = 0;
    for (size_t i = 0; i < nb; i++) {
        x = blackrock_shuffle(&blackrock,  i);
        printf("%lu\t", x);
    }
    printf("\n");
#endif
    return 0;
}
```

Then I plotted the output similarily to what you've done (regrouping values by "/10" or equivalent):

![blackblock_ivp4_100000](https://user-images.githubusercontent.com/3105926/87303299-3efc6400-c513-11ea-80e3-b8fd7f0a246f.png)

As you can see, I obtain the same result, so it seems to confirm that the issue you spotted is coming directly from the PRNG used by masscan.

**Comment by Frky** [comment]
By reading this [existing comment](https://github.com/robertdavidgraham/masscan/blob/master/src/rand-blackrock.c#L39-L42) in `rand-blackrock.c`, I did the same experiment as in my previous post with an increased number of rounds for the `blackrock` PRNG. By default, it is set to 4 (see [https://github.com/robertdavidgraham/masscan/blob/master/src/main.c#L1519](this line)), so this is the one I used before. In the following chart, I used the value 10:

![blackblock_ivp4_100000_r10](https://user-images.githubusercontent.com/3105926/87311222-40cc2480-c51f-11ea-8730-3085454a37ac.png)

You can see that the distribution is much better. Increasing the value even more is giving slightly better results, but it does not *look* way better with 20 rounds than with 10. To put a value on the "better", we should agree on a statistical metric.
For instance, here is the evolution of the standard deviation of the dataset (number of hits) depending on the number of rounds:

|  #rounds  |  standard deviation  |
|   ---   |   ---   |
|  4  |  58.677182  |
|  6  |  34.793779  |
|  8  |  20.956292  |
|  10  |  14.437481  |
|  12  |  11.536291  |
|  14  |  10.870472  |
|  16  |  9.968523  |
|  18  |  9.489151  |
|  20  |  9.827595  |
|  22  |  10.257344  |
|  24  |  9.952036  |
|  26  |  9.958516  |
|  28  |  9.412716  |
|  30  |  10.000244  |
|  32  |  9.797410  |

From these values, it seems clear that increasing the number of rounds higher than 14 is pointless. However, doubling the number of rounds (at least) increases significantly this statistical indicator. The next question is the impact on performances.

I performed the following tests: scanning 0.0.0.0/16 from a `veth` to another `veth` (which drops packets), with the following command line:

```
time sudo bin/masscan --router-mac 00:01:02:03:04:05 --adapter-ip 10.0.0.2 --range 0.0.0.0/16 --wait 0 --exclude 255.255.255.255 -p80 --rate 1000
```

Here is the result for 4 rounds:

```
Starting masscan 1.0.6 (http://bit.ly/14GZzcT) at 2020-07-13 13:37:18 GMT
 -- forced options: -sS -Pn -n --randomize-hosts -v --send-eth
Initiating SYN Stealth Scan
Scanning 65536 hosts [1 port/host]
                                                                             
real    1m6.636s
user    0m0.256s
sys     0m0.617s
```
This gave the following chart (same trend as in my previous post):

![blackblock_ivp4_100000_r4](https://user-images.githubusercontent.com/3105926/87316156-b1763f80-c525-11ea-99c2-9ce1a0e644e4.png)

And here is the result with 14 rounds:

```
Starting masscan 1.0.6 (http://bit.ly/14GZzcT) at 2020-07-13 13:56:57 GMT
 -- forced options: -sS -Pn -n --randomize-hosts -v --send-eth
Initiating SYN Stealth Scan
Scanning 65536 hosts [1 port/host]
                                                                             
real    1m6.402s
user    0m0.296s
sys     0m0.581s
```

Which gave the following chart (slight improvement of the first chart of this post):

![blackblock_ivp4_100000_r14](https://user-images.githubusercontent.com/3105926/87316181-bcc96b00-c525-11ea-89bb-f7482dd7b766.png)

My conclusion is that we could change the number of rounds for `blackrock` from 4 to 14, this would lead to a better statistical repartition during the scan, and does not seem to impact the performances significantly in my experimental setup.

**Comment by Tim---** [comment]
Hi,

Thanks for your answers ! The Zmap paper was really interesting and highlights the tradeoff between randomness and speed (Zmap uses a more expensive operation, but shards it across several threads to keep up). 

After digging a little, I also noticed that increasing the number of rounds gives better randomness (the undocumented `--blackrock-rounds` can be used for that). However, I was too lazy to search for statistical parameters to benchmark the ideal number of rounds (other than the "curve looks good" metric).

However, be careful about increasing the default number of rounds. There is zero impact on your scan because you are throttling to 1000 pps. But at "full speed", this could have a significant impact (which is probably the reason why the default number of rounds is so low). I don't have a 10gpbs connection to make the test but if anyone wants to give it a go, this could be interesting :).

**Comment by Frky** [comment]
Hi,

You're completely right, I totally forgot about the `--rate` option I used, so indeed the performance test I did has very little value... Sorry about that.

I suggest to redo the same test without the `--rate` option. As my experiment is between two `veth` one the same computer (and the receiving `veth` is just dropping packets), do you think this will give a useful measurement?

**Comment by Frky** [comment]
Thanks for your answer and your experiment.
I did the following myself, on `veth` interfaces as mentioned in my previous post:

* for `blackrock_rounds = 14`: 925741pps
```
time sudo bin/masscan --router-mac 00:01:02:03:04:05 --adapter-ip 10.0.0.2 --range 0.0.0.0/8 --wait 0 --exclude 255.255.255.255 -p80 --rate 100000000000000

Starting masscan 1.0.6 (http://bit.ly/14GZzcT) at 2020-07-20 20:40:19 GMT
 -- forced options: -sS -Pn -n --randomize-hosts -v --send-eth
Initiating SYN Stealth Scan
Scanning 16777216 hosts [1 port/host]
                                                                             
real    0m18.123s
user    0m6.658s
sys     0m9.511s
```

* for `blackrock_rounds = 4`: 1040834pps

```
time sudo bin/masscan --router-mac 00:01:02:03:04:05 --adapter-ip 10.0.0.2 --range 0.0.0.0/8 --wait 0 --exclude 255.255.255.255 -p80 --rate 100000000000000

Starting masscan 1.0.6 (http://bit.ly/14GZzcT) at 2020-07-20 20:41:07 GMT
 -- forced options: -sS -Pn -n --randomize-hosts -v --send-eth
Initiating SYN Stealth Scan
Scanning 16777216 hosts [1 port/host]
                                                                             
real    0m16.119s
user    0m4.300s
sys     0m9.754s
```

Which means an overhead, in this case, of about 12% on the time of scan.

Your point is very interesting: according to your benchmark, the overhead is much bigger but, on the other hand, it allows to send packets at more than 1Gbps which is, as you say, sufficient for most of users.

So I agree with you, it could be OK to change the default value to 14 (although, of course, the final decision goes to the owner of the project), but maybe it should be nice to also suggest a comment in the README file and/or in the command line help of `masscan` to explain this value and that it could be changed for better performances.

However, writting these lines I start to wonder whether it should not be the opposite: if the main goal of this tool is to scan fast (*i.e.*, the quality of randomness comes after), the default value should probably stay at 4...

So finally I don't really know what is the most compliant to the philosophy of the tool, although in both cases some lines in the documentation should be added. What is your view on this? (And same question to @p-l- of course.)

**Comment by mzpqnxow** [comment]
FYI, #506 seems to have been merged, you can probably close this out

**Comment by Tim---** [comment]
Oh right, thanks for the heads up.
